// @pigment-css/react/processors/createExtendSxProp.mjs
import { BaseProcessor } from './chunk-S2BUG7Y4.mjs';
import { validateParams } from '@wyw-in-js/processor-utils';

/**
  * @pigment-css/react v0.0.30
  *
  * @license MIT
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
 
var CreateExtendSxPropProcessor = class extends BaseProcessor {
  constructor(params, ...args) {
    super([params[0]], ...args);
    validateParams(params, ["callee", ["call"]], `Invalid use of ${this.tagSource.imported} tag.`);
  }
  // eslint-disable-next-line class-methods-use-this
  build() {
  }
  doEvaltimeReplacement() {
    this.replacer(this.value, false);
  }
  get value() {
    return this.astService.nullLiteral();
  }
  doRuntimeReplacement() {
    const t = this.astService;
    const extendSxPropImportIdentifier = t.addNamedImport(
      this.tagSource.imported,
      this.getImportPath()
    );
    this.replacer(t.callExpression(extendSxPropImportIdentifier, []), true);
  }
  get asSelector() {
    return `.${this.className}`;
  }
};

export { CreateExtendSxPropProcessor };
