// @pigment-css/react/build/chunk-AFZBAV6Q.mjs
import { __objRest, __spreadProps, __spreadValues } from './chunk-L25ZX2YK.mjs';
import * as React from 'react';
import clsx from 'clsx';
import isPropValid from '@emotion/is-prop-valid';
import { jsx } from 'react/jsx-runtime';

/**
  * @pigment-css/react v0.0.30
  *
  * @license MIT
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
 
function getVariantClasses(_a, variants) {
  var _b = _a, { ownerState = {} } = _b, componentProps = __objRest(_b, ["ownerState"]);
  const variantClasses = variants.filter(
    ({ props: variantProps }) => typeof variantProps === "function" ? variantProps(__spreadProps(__spreadValues(__spreadValues({}, componentProps), ownerState), { ownerState })) : Object.entries(variantProps).every(([propKey, propValue]) => {
      return ownerState[propKey] === propValue || componentProps[propKey] === propValue;
    })
  ).map(({ className }) => className);
  return variantClasses;
}
function isHtmlTag(tag) {
  return typeof tag === "string" && // 96 is one less than the char code
  // for "a" so this is checking that
  // it's a lowercase character
  tag.charCodeAt(0) > 96;
}
var slotShouldForwardProp = (key) => key !== "as" && key !== "ownerState";
var rootShouldForwardProp = (key) => slotShouldForwardProp(key) && key !== "classes";
function styled(tag, componentMeta = {}) {
  const { name, slot, shouldForwardProp } = componentMeta;
  let finalShouldForwardProp = shouldForwardProp;
  if (!shouldForwardProp) {
    if (isHtmlTag(tag)) {
      finalShouldForwardProp = isPropValid;
    } else if (slot === "Root" || slot === "root") {
      finalShouldForwardProp = rootShouldForwardProp;
    } else {
      finalShouldForwardProp = slotShouldForwardProp;
    }
  }
  let shouldUseAs = !finalShouldForwardProp("as");
  if (typeof tag !== "string" && tag.__styled_by_pigment_css) {
    shouldUseAs = false;
  }
  function scopedStyledWithOptions(options = {}) {
    const { displayName, classes = [], vars: cssVars = {}, variants = [] } = options;
    const StyledComponent = React.forwardRef(function StyledComponent2(inProps, ref) {
      const _a = inProps, { className, sx, style, ownerState } = _a, props = __objRest(_a, ["className", "sx", "style", "ownerState"]);
      const Component = shouldUseAs && inProps.as || tag;
      const varStyles = Object.entries(cssVars).reduce(
        (acc, [cssVariable, [variableFunction, isUnitLess]]) => {
          const value = variableFunction(inProps);
          if (typeof value === "undefined") {
            return acc;
          }
          if (typeof value === "string" || isUnitLess) {
            acc[`--${cssVariable}`] = value;
          } else {
            acc[`--${cssVariable}`] = `${value}px`;
          }
          return acc;
        },
        {}
      );
      const finalClassName = clsx(classes, className, getVariantClasses(inProps, variants));
      if (inProps.as && !shouldForwardProp) {
        if (!isHtmlTag(Component)) {
          if (slot === "Root" || slot === "root") {
            finalShouldForwardProp = rootShouldForwardProp;
          } else {
            finalShouldForwardProp = slotShouldForwardProp;
          }
        }
      }
      const newProps = {};
      for (const key in props) {
        if (shouldUseAs && key === "as") {
          continue;
        }
        if (finalShouldForwardProp(key) || !shouldUseAs && key === "as") {
          newProps[key] = props[key];
        }
      }
      return /* @__PURE__ */ jsx(
        Component,
        __spreadProps(__spreadValues(__spreadValues({}, newProps), Component.__styled_by_pigment_css && { ownerState }), {
          ref,
          className: finalClassName,
          style: __spreadValues(__spreadValues({}, varStyles), style)
        })
      );
    });
    let componentName = displayName;
    if (!componentName && name) {
      componentName = `${name}${slot ? `-${slot}` : ""}`;
    }
    StyledComponent.displayName = `Styled(${componentName})`;
    StyledComponent.__styled_by_pigment_css = true;
    return StyledComponent;
  }
  return scopedStyledWithOptions;
}

export { styled };
