// @pigment-css/react/build/Hidden.js
'use strict';

var chunkBV2JNSJM_js = require('./chunk-BV2JNSJM.js');
var chunkQMF3JN35_js = require('./chunk-QMF3JN35.js');
var React = require('react');
var clsx = require('clsx');
var PropTypes = require('prop-types');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
  * @pigment-css/react v0.0.30
  *
  * @license MIT
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
 
var hiddenAtomics = chunkBV2JNSJM_js.generateAtomics(({ theme }) => {
  const conditions = {};
  for (let i = 0; i < theme.breakpoints.keys.length; i += 1) {
    const breakpoint = theme.breakpoints.keys[i];
    conditions[`${theme.breakpoints.keys[i]}Only`] = theme.breakpoints.only(breakpoint);
    conditions[`${theme.breakpoints.keys[i]}Up`] = theme.breakpoints.up(breakpoint);
    conditions[`${theme.breakpoints.keys[i]}Down`] = theme.breakpoints.down(breakpoint);
  }
  return {
    conditions,
    properties: {
      display: ["none"]
    }
  };
});
var Hidden = React__namespace.forwardRef(function Hidden2(_a, ref) {
  var _b = _a, { className, component = "div", style } = _b, props = chunkQMF3JN35_js.__objRest(_b, ["className", "component", "style"]);
  const other = {};
  const breakpointProps = {};
  Object.keys(props).forEach((key) => {
    if (key.endsWith("Up") || key.endsWith("Down")) {
      breakpointProps[key] = "none";
    } else if (key === "only") {
      if (typeof props[key] === "string") {
        breakpointProps[`${props[key]}Only`] = "none";
      }
      if (Array.isArray(props[key])) {
        props[key].forEach((val) => {
          breakpointProps[`${val}Only`] = "none";
        });
      }
    } else {
      other[key] = props[key];
    }
  });
  const stackClasses = hiddenAtomics({ display: breakpointProps });
  const Component = component;
  return /* @__PURE__ */ jsxRuntime.jsx(
    Component,
    chunkQMF3JN35_js.__spreadValues({
      ref,
      className: clsx__default.default(stackClasses.className, className),
      style: chunkQMF3JN35_js.__spreadValues(chunkQMF3JN35_js.__spreadValues({}, style), stackClasses.style)
    }, other)
  );
});
if (process.env.NODE_ENV !== "production") {
  Hidden.propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes__default.default.node,
    /**
     * @ignore
     */
    className: PropTypes__default.default.string,
    /**
     * The component used for the root node.
     * Either a string to use a HTML element or a component.
     */
    component: PropTypes__default.default.elementType,
    /**
     * If `true`, screens this size and down are hidden.
     */
    lgDown: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and up are hidden.
     */
    lgUp: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and down are hidden.
     */
    mdDown: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and up are hidden.
     */
    mdUp: PropTypes__default.default.bool,
    /**
     * Hide the given breakpoint(s).
     */
    only: PropTypes__default.default.oneOfType([
      PropTypes__default.default.oneOf(["xs", "sm", "md", "lg", "xl"]),
      PropTypes__default.default.arrayOf(PropTypes__default.default.oneOf(["xs", "sm", "md", "lg", "xl"]))
    ]),
    /**
     * If `true`, screens this size and down are hidden.
     */
    smDown: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and up are hidden.
     */
    smUp: PropTypes__default.default.bool,
    /**
     * @ignore
     */
    style: PropTypes__default.default.object,
    /**
     * If `true`, screens this size and down are hidden.
     */
    xlDown: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and up are hidden.
     */
    xlUp: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and down are hidden.
     */
    xsDown: PropTypes__default.default.bool,
    /**
     * If `true`, screens this size and up are hidden.
     */
    xsUp: PropTypes__default.default.bool
  };
}
var Hidden_default = Hidden;

module.exports = Hidden_default;

module.exports = exports.default;
